### PR TITLE
[Autoscaling] Introduce resource recommenders

### LINKED ISSUE #4469: Autoscaling: handling unforeseeable volume capacities
This issue is a sub-issue of https://github.com/elastic/cloud-on-k8s/issues/4459 to focus on/discuss the case where the capacity of a persistent volume is larger than the one specified in the claim.

As an example here is a claim of `1Gi` bounded to a volume with an actual physical capacity of `368Gi`:

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    pv.kubernetes.io/bind-completed: "yes"
    pv.kubernetes.io/bound-by-controller: "yes"
  creationTimestamp: "2021-04-29T11:44:01Z"
  finalizers:
  - kubernetes.io/pvc-protection
  labels:
    common.k8s.elastic.co/type: elasticsearch
    elasticsearch.k8s.elastic.co/cluster-name: storage-sample
    elasticsearch.k8s.elastic.co/statefulset-name: storage-sample-es-data
  name: elasticsearch-data-storage-sample-es-data-0
  namespace: demo
  ownerReferences:
  - apiVersion: elasticsearch.k8s.elastic.co/v1
    kind: Elasticsearch
    name: storage-sample
    uid: b0b9b12d-092d-4c9d-a73e-8c5039bfb2d9
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi # Claim is 1Gi
  storageClassName: e2e-default
  volumeMode: Filesystem
  volumeName: local-pv-6df90d02
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 368Gi # Storage capacity is actually 1Gi
  phase: Bound
```

Having a capacity larger than the request may lead to the following situations:

1. Elasticsearch reports the total, observed, capacity as the required capacity for a tier. If the actual capacity is higher than the claim it can lead to some cascading scale up events, up to the limit specified by the user. It can also exceed to limit specified by the user in which case some not pertinent `HorizontalScalingLimitReached` events are generated.
2. If the actual capacity of a volume is greater than the claim, then the nodes may hold more data than the maximum one specified in the autoscaling specification. It may lead to overloaded nodes. For example, assuming the following autoscaling policy:
```json
{
    "name": "data",
    "roles": ["data", "ingest", "transform"],
    "resources": {
        "nodeCount": { "min": 2, "max": 5 },
        "memory": { "min": "2Gi", "max": "6Gi" },
        "storage": { "min": "1Gi",  "max": "3Gi" }
    }
}
```
Say that the claims of 1Gi have been bound to volumes of 1Ti of data each, then chances are that the 2Gi of memory are not enough to handle that amount of data.

Unforeseeable storage capacity makes it hard to scale vertically. It is especially true as long as there is no memory requirement in the autoscaling Elasticsearch API and as long as the operator attempts to infer memory requirement from storage (see https://github.com/elastic/cloud-on-k8s/issues/4076).

### Documentation update

It should be clearly stated in the documentation that scaling vertically data nodes assumes that the storage provider provisions physical volume with a predictable and even capacity across all the volumes managed by an autoscaling policy.

### Autoscaling controller update

If the operator detects that the capacity of a volume is greater than the one specified in the claim then:
  * Memory is computed according to the capacity of the volume, as reported in the PVC status, up to the memory limit specified by the user.
  * Print a log, update the status and emit a K8S warning.

### PR DESCRIPTION
**Pull request by barkbay**
This PR is mostly a refactoring of _(a part of)_ the current autoscaling code. As we want to treat the `Storage` resource in a special way this PR introduces the concept of `Recommender`.
A `Recommender` encapsulates the logic to handle requirements for a specific resource (`storage` and `memory` for now). This allows to bring the following improvements:
* Only scale storage when it is really required, this is to avoid to infinitely scale when the physical capacity of a volume is larger than the claim _(because Elasticsearch data deciders return at least the observed capacity as the required one)_.
* If the following 2 conditions are met, generate a message and cluster wide event to state that vertical autoscaling is deprecated because the node storage can't be foreseen and controlled by the operator:
  1.  the physical capacity of a volume _(as reported by ES)_ is larger than the claim
  2.  user wants to scale vertically

* Account for the 5% reserved blocks on some file systems (mostly `ext` fs): the operator always adds 5% to the storage required by Elasticsearch (AFAIK it is not possible to detect the fs type, this 5% margin then applies to any type of filesystem). It helps to:
  * calculate better storage claims or additional number of nodes to fulfil the requirements
  * spend less time in "low disk watermark" situations because the current storage capacity is not fully available to ES

Fixes #4469 and #4459 _(but still requires a documentation update to make it clear that vertical autoscaling is not supported in some conditions, to be done in a dedicated PR)_

### CODE REVIEW

**Review by pebrc — COMMENTED** [review]
LGTM. Nice refactoring I found it easy to follow 👍 Only thing we should change imo is the wording in the new event message, the rest is optional/nits. I have not tested it.

File: pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/storage.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+	currentClaimedStorage := currentResources.GetRequest(corev1.ResourceStorage)
+	if s.base.autoscalingSpec.StorageRange.Min.Equal(s.base.autoscalingSpec.StorageRange.Max) &&
+		s.base.autoscalingSpec.StorageRange.Min.Equal(currentClaimedStorage) {
+		// User provided a singular range (min == max) and the current claimed capacity is already set to the correct value, no need to scale vertically.
+		return false
+	}
+	if s.observedNodeStorageCapacity.Value() > currentClaimedStorage.Value() {
+		// Log a warning and ensure we max out the storage in the claim to also scale up dependant resources like memory or CPU.
+		s.log.Info(
+			"Vertical Pod autoscaling deprecated: current node storage capacity is greater than the claimed capacity.",
```
[the remark this answers is not in the document]
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/storage.go:90]: You're right, "deprecated" is poorly chosen. What I would like to indicate here is that the user should not use vertical autoscaling at all. The linear algorithm used to scale cpu and memory according to the storage would be ineffective.

What about "Vertical Pod autoscaling is not supported: current node storage capacity is greater than the claimed capacity" ?

[the remark this answers is not in the document]
Comment (pebrc) [reply pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/storage.go:90]: > What about "Vertical Pod autoscaling is not supported: current node storage capacity is greater than the claimed capacity" ?

I think that works 👍

**Review by pebrc — APPROVED** [review]
LGTM 👍

File: pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/storage.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+	// * Elasticsearch API returned a non zero requirement
+	if !hasResourceRange && !hasRequirement {
+		return &nilRecommender{}, nil
+	}
+
+	// Ensure that we have all the information needed to recommend something.
+	// In case of storage it means that we must have the observed node capacity.
+	if autoscalingPolicyResult.CurrentCapacity.Total.Storage == nil ||
+		autoscalingPolicyResult.CurrentCapacity.Node.Storage == nil {
+		return nil, fmt.Errorf("observed storage capacity is expected in Elasticsearch autoscaling API response")
```
[the remark this answers is not in the document]
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/autoscaler/recommender/storage.go:159]: The observed capacity is also used to detect when the current capacity is greater than the claimed one and alert the user. I share your concern but I think it is an important information required to make the right decisions.
I added a [specific step in the e2e test](https://github.com/elastic/cloud-on-k8s/commit/24330cfeee5c6b3579ed64980fa2236b50c7d727) to early detect such a change in the Elasticsearch API. If it turns out to be the case at some point we can revisit that choice to find an alternative ?

**Pull request merged**